### PR TITLE
 snapstate: drop revisions after "current" on refresh

### PR DESCRIPTION
**Pull request by mvo5**
Based on https://github.com/snapcore/snapd/pull/1275#discussion_r68764940

For easy review use: https://github.com/mvo5/snappy/compare/feature/rollback...mvo5:feature/rollback2?expand=1

### CODE REVIEW

**Comment by niemeyer** [comment]
LGTM

**Comment by niemeyer** [comment]
Needs conflict solving, and a second one from you @pedronis would be good as you're in context.

**Comment by niemeyer** [comment]
Tests are broken due to the OfficialName => RealName change.

File: overlord/snapstate/snapstate.go [hunk 1]
```diff
@@ -119,11 +119,23 @@ func doInstall(s *state.State, snapst *SnapState, ss *SnapSetup) (*state.TaskSet
 	linkSnap := s.NewTask("link-snap", fmt.Sprintf(i18n.G("Make snap %q available to the system%s"), ss.Name, revisionStr))
 	addTask(linkSnap)
 
-	// do GC!
-	if snapst.HasCurrent() {
+	// Do not do that if we are reverting to a local revision
+	if ss.Revision.Unset() && snapst.HasCurrent() {
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go]: I think this check is no longer going to be right in the new world (Revision is sometimes going to be set for installs as well, need to be !revisionIsLocal ? )

File: overlord/snapstate/snapstate.go [hunk 2]
```diff
@@ -121,11 +121,23 @@ func doInstall(s *state.State, snapst *SnapState, ss *SnapSetup) (*state.TaskSet
 	linkSnap := s.NewTask("link-snap", fmt.Sprintf(i18n.G("Make snap %q%s available to the system"), ss.Name(), revisionStr))
 	addTask(linkSnap)
 
-	// do GC!
-	if snapst.HasCurrent() {
+	// Do not do that if we are reverting to a local revision
+	if snapst.HasCurrent() && !revisionIsLocal {
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go]: does this now needs to be !isRevert or this still what we want?

Comment (mvo5) [inline overlord/snapstate/snapstate.go]: I think this is still what we want (unless I'm missing something of course). I think we should not mess around with garbage-collection or discarding version as long as you jump around local revisions. This code (gc/discard-after-current) will run naturally once the snap is refreshed again to a store version (or when it gets removed).

**Comment by pedronis** [comment]
retest this please

**Comment by mvo5** [comment]
Closing for now as we discussed a new approach for revert and refresh.

**Comment by mvo5** [comment]
retest this please

**Comment by mvo5** [comment]
We still need this one with the latest approach

**Comment by chipaca** [comment]
:+1:

**Pull request merged**